### PR TITLE
[TRTLLM-11064][fix] Remove duplicated MoE Computation with Helix CP+DP

### PR DESCRIPTION
**Pull request by brb-nv**
### Background

When using Helix CP with Attention DP on the generation server, all CP ranks within the same DP group have identical data after attention computation. The MoE layer receives token counts from all ranks via `tp_cp_allgather` to coordinate expert routing and communication.

### Issue

MoE computation was being duplicated across CP ranks. Each CP rank was processing the same tokens independently, resulting in redundant computation and poor perf while maintaining accuracy.

### Root Cause

The `tp_cp_allgather` operation gathers token counts from all TP×CP ranks. In Helix mode, CP ranks hold identical data post-attention, so the gathered token counts included duplicates. For example, with CP=2:
- `[dp0cp0=128, dp0cp1=128, dp1cp0=64, dp1cp1=64]`

The MoE layer would then process 128 tokens on both `dp0cp0` and `dp0cp1`, doubling the computation.

### Fix

Allgather at the beginning of MLA layer and ReduceScatter at the end of it when AttentionDP is enabled.

## Test Coverage
```
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[fifo_v2-cudagraph:with_padding-pp1dp2cp2] -s -v
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Release Notes

* **New Features**
  * Added support for Helix Context Parallelism in DeepseekV3 models, enabling optimized distributed inference with automatic token deduplication across compute ranks.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by syuoni — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -1451,9 +1399,58 @@ def _postprocess_inputs(self, inputs: Dict[str, Any]):
 
     def _get_all_rank_num_tokens(self, attn_metadata: AttentionMetadata):
         if self.enable_attention_dp:
-            return list(self.dist.tp_cp_allgather(attn_metadata.num_tokens))
+            all_rank_num_tokens = list(
+                self.dist.tp_cp_allgather(attn_metadata.num_tokens))
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Can we use `tp_allgather` instead of `tp_cp_allgather` to avoid duplication?

Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Hi Yuxian, thanks to your detailed feedback, changes are much cleaner and limited to MLA module. However, `tp_cp_allgather` is needed.

- There's a mapping mismatch between `model_engine` and `MoE`. `model_engine` uses the original mapping (e.g., tp_size=1, cp_size=8). MoE modules use the repurposed mapping (e.g., tp_size=8, cp_size=1).
In MoE, `self.parallel_rank = self.mapping.tp_rank` uses the repurposed `tp_rank` (0–7), and all these sites index into `all_rank_num_tokens[rank]`:
- [interface.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/modules/fused_moe/interface.py#L713-L714)
- [fused_moe_cutlass.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py#L1020)
- [fused_moe_deepgemm.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/modules/fused_moe/fused_moe_deepgemm.py#L956)
[fused_moe_trtllm_gen.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py#L1042)
[configurable_moe.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/modules/fused_moe/configurable_moe.py#L467)
[fused_moe_wide_ep.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py#L845)

So, using tp_allgather would cause OOB lookup. Please let me know if you're ok with this. Alternative is pass original `mapping_with_cp` to MoE also and update all usage to account for CP. For example, lookup should be with `self.parallel_rank // cp_size`.

File: tensorrt_llm/_torch/models/modeling_speculative.py [hunk 2]
```diff
@@ -236,11 +238,19 @@ def forward(
             embeds = self.input_layernorm(embeds)
             hidden_states = torch.cat([embeds, hidden_states], dim=-1)
 
-        hidden_states = self.self_attn(
-            position_ids=position_ids,
-            hidden_states=hidden_states,
-            attn_metadata=attn_metadata,
-        )
+        if self._use_mla:
+            hidden_states, residual = self.self_attn(
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/modeling_speculative.py]: We only handle `residual` for the first layer, so it is unrelated to `Eagle3DecoderLayer`. We can update MLA to support both w/ residual and w/o residual cases like https://github.com/NVIDIA/TensorRT-LLM/blob/f414a079ad8497ce4399e6677598daf716985e23/tensorrt_llm/_torch/modules/decoder_layer.py#L31
so that we don't need to change this file at all.

File: tensorrt_llm/_torch/modules/attention.py [hunk 3]
```diff
@@ -2278,14 +2280,108 @@ def forward_sparse_mla_kvcache_bf16(
...
+        start = self.mapping.cp_rank * chunk_size
+        return tensor[start:start + chunk_size]
+
+    def _output_projection(
+        self,
+        attn_output: torch.Tensor,
+        attn_metadata: AttentionMetadata,
+        all_reduce_params: Optional[AllReduceParams],
+        residual: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py]: We prefer to use `tuple` instead of `Tuple` now, can you update all 3 usages of `Tuple`?

Comment (brb-nv) [reply tensorrt_llm/_torch/modules/attention.py]: Updated. https://github.com/NVIDIA/TensorRT-LLM/pull/11167/commits/5d5a5942f2cf73d11ea1a27a9e9bb722b385e622

File: tensorrt_llm/_torch/modules/attention.py [hunk 4]
```diff
@@ -2278,14 +2280,110 @@ def forward_sparse_mla_kvcache_bf16(
...
+            attn_output, _ = self._pad_for_cp(attn_output,
+                                              attn_metadata.num_tokens)
+
+            # Reduce-scatter using mapping_o where tp_group = cp_group.
+            attn_output = reducescatter(attn_output, self.mapping_o, dim=0)
+
+            # For the first layer, the residual comes from the embedding and
+            # has not been through a prior RS. Slice it to match.
+            if self.layer_idx == 0 and residual is not ...:
+                residual = self._slice_for_cp(residual, attn_metadata)
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py]: Can we move this part out of MLA and rename it to `maybe_slice_for_cp` so that we can directly call it in `modeling_deepseekv3.py`? It can be reused between `MLA` and `Attention`. By calling it explicitly, we can avoid adding `residual` to MLA's forward, which keeps the interface unchanged.

**Review by yuxianq — APPROVED** [review]

File: tensorrt_llm/_torch/modules/attention.py (hunk 4 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/modules/attention.py]: Will take care of this in a follow-up MR as discussed offline.
